sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala

```diff
@@ -0,0 +1,112 @@
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/JoinBenchmark-results.txt".
+ * }}}
+ */
+
+object CollationBenchmark extends SqlBasedBenchmark {
+  private val collationTypes = Seq("UTF8_BINARY", "UTF8_BINARY_LCASE", "UNICODE", "UNICODE_CI")
+
+  def generateUTF8Strings(n: Int): Seq[UTF8String] = {
+    // Generate n UTF8Strings
+    Seq("ABC", "aBC", "abc", "DEF", "def", "GHI", "ghi", "JKL", "jkl",
+      "MNO", "mno", "PQR", "pqr", "STU", "stu", "VWX", "vwx", "YZ").map(UTF8String.fromString) ++
+    (18 to n).map(i => UTF8String.fromString(i.toOctalString))
+  }
+
+  def benchmarkUTFString(collationTypes: Seq[String], utf8Strings: Seq[UTF8String]): Unit = {
+    val sublistStrings = utf8Strings
+    val benchmark = collationTypes.foldLeft(
+      new Benchmark(s"collation unit benchmarks", utf8Strings.size, output = output)) {
+      (b, collationType) =>
+        val collation = CollationFactory.fetchCollation(collationType)
+        b.addCase(s"equalsFunction - $collationType") { _ =>
+          sublistStrings.foreach(s1 =>
+            utf8Strings.foreach(s =>
+            collation.equalsFunction(s, s1).booleanValue()
+            )
+          )
+        }
+        b.addCase(s"collator.compare - $collationType") { _ =>
+          sublistStrings.foreach(s1 =>
+            utf8Strings.foreach(s =>
+              collation.comparator.compare(s, s1)
+            )
+          )
+        }
+        b.addCase(s"hashFunction - $collationType") { _ =>
+          sublistStrings.foreach(_ =>
+            utf8Strings.foreach(s =>
+              collation.hashFunction.applyAsLong(s)
+            )
+          )
+        }
+        b
+    }
+    benchmark.run()
+  }
+
+  def df1: DataFrame = {
+    val d = spark.createDataFrame(Seq(
+      ("ABC", "ABC"), ("aBC", "abc"), ("abc", "ABC"), ("DEF", "DEF"), ("def", "DEF"),
+      ("GHI", "GHI"), ("ghi", "IG"), ("JKL", "NOP"), ("jkl", "LKJ"), ("mnO", "MNO"),
+      ("hello", "hola"))
+    ).toDF("s1", "s2")
+    d
+  }
+
+  def collationBenchmarkFilterEqual(collationTypes: Seq[String]): Unit = {
+    val benchmark = collationTypes.foldLeft(
+      new Benchmark(s"filter df column with collation", 11, output = output)) {
+      (b, collationType) =>
+        b.addCase(s"filter df column with collation - $collationType") { _ =>
+          val df = df1.selectExpr(
+            s"collate(s2, '$collationType') as k2_$collationType",
+            s"collate(s1, '$collationType') as k1_$collationType")
+
+          df.where(col(s"k1_$collationType") === col(s"k2_$collationType"))
+                .queryExecution.executedPlan.executeCollect()
+
+        }
+        b
+    }
+    benchmark.run()
+  }
+
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    collationBenchmarkFilterEqual(collationTypes.reverse)
```

Comment:
Why doing `collationTypes.reverse` on collationBenchmarkFilterEqual and not in `benchmarkUTF8String`?

We will probably be adding new benchmarks in the future so lets leave clear format here.
1) All benchmarks should run against source that is provided in consistent way. Here `collationBenchmarkFilterEqual` generates data on it's own, while `benchmarkUTF8String` accepts data through `generateUTF8Strings`. I don't understand reason for this.

2) Nit: Naming should be consistent. E.g. both should be `benchmarkFilterEqual` and `benchmarkUTFString`.

Reply:
Good points. Let me know if you are happy with my changes, which I believe incorporated this feedback.